DRILL-8287: Add Support for Keyset Based Pagination

cgivre opened the pull request:
# [DRILL-8287](https://issues.apache.org/jira/browse/DRILL-8287): Add Support for Keyset Based Pagination

## Description
Some APIs such as HubSpot use values in the result set to indicate whether there are additional pages.  This PR adds support for this kind of pagination.  Note that current implementation only works for JSON based APIs.

This PR also addresses [DRILL-8286](https://issues.apache.org/jira/browse/DRILL-8286), which is a minor bugfix for the GoogleSheets config. 

## Documentation
Updated Pagination.md.

## Testing
Added unit tests and manually tested against Hubspot API.

jnturton commented:
I'm not sure that the concept of pagination from the HTTP plugin should spill into the JSON reader. Can you abstract it, e.g. by renaming paginationMap to, say, listenerColumnMap?

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpPaginatorConfig.java:
```diff
@@ -137,21 +162,28 @@ public String toString() {
   }
 
-  private HttpPaginatorConfig(HttpPaginatorConfig.HttpPaginatorBuilder builder) {
+  /*private HttpPaginatorConfig(HttpPaginatorConfig.HttpPaginatorConfigBuilder builder) {
```
Oops... Fixed.

cgivre commented:
> I'm not sure that the concept of pagination from the HTTP plugin should spill into the JSON reader. Can you abstract it, e.g. by renaming paginationMap to, say, listenerColumnMap?

Fixed.

cgivre commented:
@jnturton Thanks for the quick review!  I addressed your comments.   I actually reinserted the commented out block as that was intended to make sure that the user properly populates the pagination fields.  Not sure why I commented that out in the first place.

jnturton commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/parser/SimpleMessageParser.java:
```diff
@@ -66,11 +68,13 @@
+  private final Map<String, Object> paginationFields;
 
-  public SimpleMessageParser(String dataPath) {
+  public SimpleMessageParser(String dataPath, Map<String, Object> paginationFields) {
```
Can we rename "pagination" here too?

cgivre replied:
Done!

jnturton commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/values/ScalarListener.java:
```diff
@@ -76,4 +79,33 @@ protected void setArrayNull() {
   }
+
+  /**
+   * Adds a field's most recent value to the pagination map.  This is necessary for the HTTP plugin
+   * for index or keyset pagination where the API transmits values in the results that are used to
+   * generate the next page.
+   *
+   * This data is only stored if the pagination map is defined, and has keys.
```
Can this be rewritten in terms of generic column listeners rather than pagination and the HTTP plugin?

cgivre replied:
Done!

jnturton commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/loader/TupleParser.java:
```diff
@@ -127,10 +127,19 @@ public TupleParser(JsonLoaderImpl loader, TupleWriter tupleWriter, TupleMetadata
+  }
+
+  private boolean projectField(String key) {
+    // This method makes sure that fields necessary for pagination are read.
```
```suggestion
    // This method makes sure that fields necessary for column listeners are read.
```

jnturton commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/values/ScalarListener.java:
```diff
@@ -76,4 +79,30 @@ protected void setArrayNull() {
+  }
+
+  protected void addValueToListenerMap(String key, Object value) {
+    Map<String, Object> paginationMap = loader.listenerColumnMap();
```
```suggestion
    Map<String, Object> listenerMap = loader.listenerColumnMap();
```

jnturton commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/parser/SimpleMessageParser.java:
```diff
@@ -129,6 +135,44 @@ private boolean parseInnerLevel(TokenIterator tokenizer, int level) throws Messa
+   * without creating a ValueVector for them.
+   *
+   * @param tokenizer A {@link TokenIterator} of the parsed JSON data.
+   * @param fieldName A {@link String} of the pagination field name.
```
```suggestion
   * @param fieldName A {@link String} of the listener column name.
```

jnturton commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/values/ScalarListener.java:
```diff
@@ -76,4 +79,30 @@ protected void setArrayNull() {
+  /**
+   * Adds a field's most recent value to the column listener map.
+   * This data is only stored if the listener column map is defined, and has keys.
+   * @param key The key of the pagination field
```
```suggestion
   * @param key The key of the listener field
```

jnturton reviewed (approved):
Thanks for making changes.
